trie/fog.py

```diff
@@ -0,0 +1,284 @@
+import ast
+from itertools import zip_longest
+from typing import (
+    Any,
+    Sequence,
+    Tuple,
+)
+
+from eth_utils import (
+    to_tuple,
+    ValidationError,
+)
+from sortedcontainers import SortedSet
+
+from trie.typing import (
+    Nibbles,
+    NibblesInput,
+    RawHexaryNode,
+)
+from trie.utils.nodes import (
+    key_starts_with,
+)
+
+
+class HexaryTrieFog:
+    """
+    Keeps track of which parts of a trie have been verified to exist.
+
+    Named after "fog of war" popular in video games like... Red Alert? IDK, I'm old.
+
+    Object is immutable. Any changes, like marking a key prefix as complete, will
+    return a new HexaryTrieFog object.
+    """
+    _unexplored_prefixes: SortedSet
```

Comment:
Depends on how well `sortedcontainers` does their typing, but this might be best typed as `typing.Set[...]` so that the object's items are also typed.

Reply:
We use APIs that aren't in `Set` like indexing to a value and `.bisect()`. I thought I might find a quick solution, then ended up building a `Protocol`, oops.